### ISSUE TITLE
assemble syntax problem

### ISSUE THREAD

**Issue by xubenji** [nneonneo/ghidra-wasm-plugin#19]
Hello, When I try to assemble WASM instruction like this:
```
ins = "local.get 88"
asm.assemble(currentAddress, ins)
```
I will get this error in the ghidra command:
<img width="1295" alt="截屏2024-02-03 下午2 41 37" src="https://github.com/nneonneo/ghidra-wasm-plugin/assets/33605135/2e646b92-6b0f-420d-88a2-23ca2f9f95ed">
It looks like a syntax problem. I don't think the plugin can support the WASM very perfectly or it is my mistake. What is the correct WASM instruction syntax?

**Comment by nneonneo** [comment]
Unfortunately, assembling or patching isn’t supported. There’s a bunch of reasons for this, but basically it boils down to the plugin being unable to perform the reanalysis necessary to handle modified code. In addition, as you’ve noticed, the LEB128 parsing that we’re using for the instructions doesn’t invert properly in SLEIGH, although that could be worked around if someone wanted to try fixing that. I would recommend using something like wasm2wat + wat2wasm to convert the module into a textual form, patch it, and convert back.RobertOn Feb 3, 2024, at 11:44 AM, Benji Xu ***@***.***> wrote:﻿
Hello, When I try to assemble WASM instruction like this:
ins = "local.get 88"
asm.assemble(currentAddress, ins)

I will get this error in the ghidra command:
2024-02-03.2.41.37.png (view on web)
It looks like a syntax problem. I don't think the plugin can support the WASM very perfectly or it is my mistake. What is the correct WASM instruction syntax?

**Comment by xubenji** [comment]
looks like the plugin can't correctly recognize the basic block of webassembly bytecode.
The definition of basic block in wasm is start from `block` end at `end`. But when I use:
```
    for block in high_function.getBasicBlocks():
        print(block)
```
The terminal print the wrong basic block start and end address.

**Comment by nneonneo** [comment]
You can use an external tool like wasm2wat/wat2wasm from [wabt](https://github.com/WebAssembly/wabt) for that. There is even a web version if you don't like downloading extra software: https://webassembly.github.io/wabt/demo/wasm2wat/

**Comment by xubenji** [comment]
I just want to find out if we can re-write bytecode based on opcodeops. For example, if I insert a certain opcode in opcodeops, will the ghidra automatically re-write the wasm bytecode?

**Comment by nneonneo** [comment]
No, unfortunately not. Wasm bytecodes are variable length (and encoded using a variable-length-integer scheme), so even small changes may cause the instructions to change length, which might necessitate rewriting the entire section.

The plugin also uses some very fancy tricks in SLEIGH to decode the varint opcodes, which probably breaks the automated assembler. Making the assembler work was not a priority in development, although any ideas for how to fix this would be welcome.

Even if the bytecode is the same length after the change, the Wasm plugin doesn't currently support modification due to its design - the only way to determine the functionality of certain opcodes is to perform a full Wasm stack analysis on the function, and any change would necessitate a re-analysis. This is also technically fixable (trigger a re-analysis on any code patch), but again is not a priority for this project.
